### PR TITLE
ENH: Add option to display count of observations in `summary2.summary_col`

### PR DESCRIPTION
**Pull request by EBoisseauSierra**
- [x] Linked to #8996
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

This PR introduces a new optional argument to [`statsmodels.iolib.summary2.summary_col`](https://github.com/statsmodels/statsmodels/blob/v0.14.0/statsmodels/iolib/summary2.py#L471), which enable easily displaying the count of observations in a summary table:

```diff
  def summary_col(
      results,
      float_format: str ='%.4f',
      model_names=(),
      stars=False,
      info_dict=None,
      regressor_order=(),
      drop_omitted=False,
      include_r2=True,
+     include_n: bool =False,
  ):
```

Example:

```
=========================
                    y    
-------------------------
const            0.5000  
                 (0.6708)
x1               0.6000  
                 (0.2449)
R-squared        0.3000  
R-squared Adj.   0.2500  
No. Observations 16      
=========================
Standard errors in
parentheses.
```

The current workaround is to use `info_dict` parameter as follows:

```diff
  def summary_col(
      results,
      float_format: str ='%.4f',
      model_names=(),
      stars=False,
+     info_dict={
+       'No. Observations': lambda x: "{0:d}".format(int(x.nobs)),
+     }
      regressor_order=(),
      drop_omitted=False,
  )
 ```
 
However, I believe that easily toggling the display of the count of observations is a frequent enough use-case to add this new parameter.

### CODE REVIEW

File: statsmodels/iolib/summary2.py [hunk 1]
```diff
@@ -556,10 +569,8 @@ def merg(x, y):
     index = summ.index.get_level_values(0)
     for i in range(0, index.shape[0], 2):
         idx.append(index[i])
-        if (i + 1) < index.shape[0] and (index[i] == index[i + 1]):
-            idx.append("")
-        else:
-            idx.append(index[i + 1])
+        if (i + 1) < index.shape[0]:
+            idx.append(index[i + 1] if index[i] != index[i + 1] else "")
```
Comment (josef-pkt) [inline statsmodels/iolib/summary2.py:573]: AFAICS This changes the else path when `(i + 1) == index.shape[0]`
which version is correct?

Comment (josef-pkt) [reply statsmodels/iolib/summary2.py:573]: answer in commit "fix: Allow odd number of elements in index"

**Comment by josef-pkt** [comment]
first commit is bug fix
all other commits can be squashed AFAICS

**Comment by josef-pkt** [comment]
One possible problem here is definition of nobs with `freq_weights`.
Need to check how GLM defines nobs/wnobs in that case.

Also do all model have nobs as results attribute?

name of option? include_nobs instead of include_n?

**Comment by josef-pkt** [comment]
This PR implements nobs in the same way as R-squared.

instead I might remove the current R-squared handling and add Rsquared and nobs to info dict instead, see #9282

File: statsmodels/iolib/summary2.py (hunk 1 above)
Comment (EBoisseauSierra) [reply statsmodels/iolib/summary2.py:573]: Indeed (db364565 for ref).

Currently, if `(i+1) < index.shape[0] is False`, the `else` clause will still try to access `index[i+1]`.

**Comment by EBoisseauSierra** [comment]
TL;DR: @josef-pkt , could you please clarify how you'd like me to move this PR forward?

> first commit is bug fix, all other commits can be squashed AFAICS

Do you want me to raise a separate PR for the bug fix?
Do you want me to squash the commits or are you happy with "squash-merge" the PR (what I'd prefer)?

> this will likely have merge conflicts with https://github.com/statsmodels/statsmodels/pull/9191

I've seen its successor has been merge, I'll thus rebase this one shortly.

> Also do all model have nobs as results attribute?

Great thinking. You'd be much better informed to answer this one.

> name of option? include_nobs instead of include_n?

I'm happy with either.

> instead I might remove the current R-squared handling and add Rsquared and nobs to info dict instead

Now I'm unsure how you'd like to move this PR forward? Would you prefer me to chip in to #9282? (Using `info_dict` seems a better approach indeed.)

**Comment by josef-pkt** [comment]
@EBoisseauSierra 

Thanks for the PR.
I have now finally gone through the summary_coll code, triggered by this and related PRs.
And now I prefer to refactor the code before making additional enhancements.
see #9283

For now we keep these PRs open.
But the option for adding nobs will be implemented in a different way from the version in this PR.
And I want to similarly refactor the R2 handling.

Adding the option to choose which parameter statistic is included will most likely follow the other PR, but so far I have only skimmed the details for that.

I will soon be travelling and going on vacation, so this will likely happen in August, but I expect that it will be in the next release.
It would be helpful if you could review the changes before the merge. Especially having an api for the options that are flexible and convenient and don't fill up the signature with `include_xxx` keywords will be difficult.

Another design issue will be which statistics to include in the  default info_dict for the different models, OLS, GLM, discrete, ...

**Comment by josef-pkt** [comment]
(on vacation one year later)

what's that status for the related issues and PRs?
I have not looked at this in a year.
